### PR TITLE
fix(security): close current CodeQL high alerts

### PR DESCRIPTION
**Pull request by john-the-dev**
## What changed, and why?

Closes the four open high-severity CodeQL findings on `main` (alerts 55–58): three `py/path-injection` flows in the authenticated TaskDelegationService endpoints and one `py/insecure-temporary-file` finding in the class-rule test.

The submit route now normalizes the final task path and proves it remains inside `TASK_DIR` immediately before writing. The archive route applies the same CodeQL-modeled `realpath` + prefix check locally at its source and destination sinks. The test fixture now uses `NamedTemporaryFile` instead of deprecated `mktemp`.

## Review first

1. `src/agent-api.py` — containment checks in `delegation_submit_task` and `delegation_archive_result`.
2. `tests/agent-api-delegation.test.py` — task-write, archive-source, and archive-destination symlink regressions.
3. `tests/lint-class-rules.test.py` — atomic temporary-file creation.

## Failing-before / passing-after evidence

On `origin/main` (`b6ec180`), a pre-positioned `tasks/task-symlink-escape.txt` symlink caused a valid authenticated delegation submit to return 200 and overwrite the target outside `tasks/`. The reproduced target changed from `sentinel` to the HTTP-controlled task body.

On this head (`ead4ada`), the same case returns 400 and the external target remains unchanged. The regression is part of the delegation E2E suite.

## Tests run

- `python3 tests/agent-api-delegation.test.py` — PASS, including all three symlink-containment cases.
- `python3 tests/lint-class-rules.test.py` — PASS (31 tests).
- `python3 tests/agent-api-task-display-text.test.py` — PASS.
- `python3 -m py_compile src/agent-api.py tests/agent-api-delegation.test.py tests/lint-class-rules.test.py` — PASS.
- `git diff --check` — PASS.

GitHub CodeQL is the authoritative verification that alerts 55–58 close; its PR workflow is expected to run on this branch.

## Edge cases checked

- Lexically valid task ID with an existing symlink outside `tasks/`.
- Archive source symlink resolving outside `results/`.
- Archive destination symlink resolving outside the month directory.
- Existing archive destination still uses the timestamp-suffixed no-clobber path.
- Traversal, mismatched IDs, malformed IDs, and already-gone results retain their existing responses.

## Migration, config, permissions, rollback, and deployment risks

No migration, config, or permission changes. Existing regular files behave unchanged. A request that resolves through a symlink outside its permitted root now fails with HTTP 400. Rollback is a normal revert.

## Known gaps / follow-up

None for the current open CodeQL alert set.

### CODE REVIEW

**Review by liususan091219 — APPROVED** [review]
LGTM security: realpath+os.sep containment at all path sinks; exploit-reproduction regression tests assert 400 and outside file untouched; insecure mktemp replaced. Closes CodeQL 55-58.

File: src/agent-api.py [hunk 1]
```diff
@@ -364,21 +364,40 @@ def delegation_submit_task(data: dict):
...
+    # tainted HTTP value through normalization to the filesystem sink.  The
+    # task-id regex alone prevents lexical traversal, but not an existing
+    # `<task-id>.txt` symlink that points outside TASK_DIR.
+    task_dir_real = os.path.realpath(TASK_DIR)
+    task_file_str = os.path.realpath(
+        os.path.join(task_dir_real, f"{tid}.txt")
+    )
+    if not task_file_str.startswith(task_dir_real + os.sep):
+        return 400, {"error": "task path escapes task directory"}
+    Path(task_file_str).write_text(content)
```
Comment (qingyun-wu) [inline src/agent-api.py]: The containment check and the write are separate path resolutions. A local attacker can replace `task_file_str` with a symlink after `realpath()`/the prefix check and before `write_text()`, so the sink can still follow a link outside `TASK_DIR`. Please make the open itself no-follow/atomic (for example `os.open` with `O_NOFOLLOW` where available plus an `fstat`/directory-fd strategy), rather than validating a pathname and reopening it later.

**Review by qingyun-wu — CHANGES_REQUESTED** [review]
The focused delegation test passes, but the task-write protection still has a race that preserves the file-write vulnerability under concurrent local filesystem access.

File: src/agent-api.py (hunk 1 above)
Comment (john-the-dev) [reply src/agent-api.py]: Fixed in d7ed7f5. The submit path no longer validates and then reopens the destination pathname: it writes a private O_EXCL/O_NOFOLLOW temp entry through an open TASK_DIR descriptor and atomically publishes with os.replace using source/destination dir_fds. A raced symlink is replaced as a directory entry, never followed. Added a deterministic race regression that injects the symlink immediately inside the replace call; the outside sentinel remains unchanged and the final task is a regular file. Focused delegation E2E, 31 lint-class tests, py_compile, and diff-check all pass.

**Comment by john-the-dev** [comment]
Follow-up 5a50c97 adds focused coverage for the descriptor/publish failure paths: in-directory preexisting symlink rejection, fdopen failure cleanup, and a publish failure where the temp name is already gone. The delegation E2E and lint-class suites pass locally; refreshed CI is now 12/12 green, including diff coverage. @qingyun-wu re-review requested on the race fix.

**Review by qingyun-wu — APPROVED** [review]
The follow-up resolves the previously reported task-write race: content is written to a private O_EXCL entry through an open task-directory descriptor and published with a dir-fd-scoped atomic replace, so a raced destination symlink is replaced rather than followed. The added tests cover the deterministic symlink swap, pre-existing in-directory symlink rejection, write cleanup, and publish-failure cleanup. Focused delegation E2E, 31 lint-class tests, py_compile, and diff checks pass.
